### ISSUE TITLE
Create a class hierarchy for proxy-loss types: Quadratic/XEntropy/Hinge

### ISSUE THREAD

**Issue by AndreFCruz** [feedzai/fairgbm#9]
## Summary

- Currently we have to do string comparisons multiple times per data row, per iteration, to figure out which proxy loss function to use to compute the gradients.
  - e.g., https://github.com/feedzai/fairgbm/blob/bulk-fairgbm-modifications/src/objective/constrained_recall_objective.hpp#L86
- According to our benchmarks, this can take up around 30% of the total training time.
- We should create a common interface w/ different subclasses to optimize this.
- This has already been started in #18

**Comment by AndreFCruz** [comment]
The skeleton for this refactor has been started: I've created a common interface for all proxy losses at `include/LightGBM/utils/loss_proxies.hpp`

**Comment by AndreFCruz** [comment]
Closed by #35
